### ISSUE TITLE
_score is implicitly converted to integer with groovy

### ISSUE THREAD

**Issue by larschri** [elastic/elasticsearch#14786]
It looks like _score is implicitly converted to integer in this groovy script: ``_score * 1``. I assume this is a bug since _score contains a floating point value.

To reproduce:
```
# Create empty index
curl -XPUT 'http://localhost:9200/twitter/'

# Feed two documents
for a in doc1 doc2; do curl -XPUT http://localhost:9200/twitter/tweet/$a -d '{"foo":"foo"}'; done

# Search with script_score
curl http://localhost:9200/twitter/_search -d '{
  "query": {
    "function_score": {
      "query": {
        "term": {
          "foo": "foo"
        }
      },
      "functions": [
        {
          "script_score": {
            "script": "_score * 1"
          }
        }
      ]
    }
  }
}'
```

Notice that hits have score 0.0. Replace ``_score * 1`` by ``_score`` to get the correct score.

I am running elasticsearch 1.7.3 and groovy 2.4.4

**Comment by clintongormley** [comment]
A few interesting things here:

*  `_score * 1` is resulting in an integer, which can result in a score of 0 if the score is less than 1
* `_score * 10000` STILL results in a score of 0
* `_score * 1.0` works
* In plain groovy, `_score = 0.94; _score * 1` returns a floating point

@dakrone any ideas what is happening here?

**Comment by larschri** [comment]
``_score * 10000`` - My theory is that _score is casted to integer (zero) before the multiplication takes place.

**Comment by clintongormley** [comment]
ah, thanks for the explanation @rjernst.   I don't think this is anything we can fix? So the solution here is to use `1.0` instead.

**Comment by larschri** [comment]
> ah, thanks for the explanation @rjernst. I don't think this is anything we can fix? So the solution here is to use 1.0 instead.

Using ``1.0`` is fine when you are aware of the bug, but the bug is still there. Users expect ``_score`` to behave like a float, so why not use ``float`` instead of ``ScoreAccessor``? If that's not possible, maybe it works to inherit from ``java.math.BigDecimal`` instead of ``Number``? Elasticsearch could even start every groovy script with ``_score = (float) _score;`` to fix this.

BTW. This looks like the same problem http://stackoverflow.com/questions/30048963/groovy-numbermath-defaults-to-integermath-for-unkown-number-types
